### PR TITLE
GPflow 2.0: ChangePoints

### PR DESCRIPTION
**Pull request by jahall**
Introduction of a `ChangePoints` kernel to allow the identification of regime changes in 1d functions or timeseries. Examples provided in the associated notebook. I would welcome feedback on the design, and whether this would be something useful to have in the core library.

This is a port of previous PR https://github.com/GPflow/GPflow/pull/1065 to gpflow2.

### CODE REVIEW

**Comment by jahall** [comment]
> You can also change the merge target for Pull Requests - no need to close and open a new one just to change that it should be merging into `develop-2.0` instead of `develop`. :)

Sorry about this - I was a little hasty in deleting and re-opening!

**Comment by jahall** [comment]
@frgsimpson as per your comments from the previous PR:
- _Explanation of np.log(0.95)_ - I have simplified this so that the steepness parameter of the sigmoid can be passed in directly so things are a little clearer
- _Multiple kernels tests_ - Done
- _Move to pytest_ - Done
- _Hierarchical flattening of changepoints_ - I overrode the flattening behaviour for changepoints as it is not clear how to flatten in a sensible manner
- _Activation order_ - I have removed this in favour of direct specification via `kernels`
- _Efficient `K_diag`_ - Still TODO

**Comment by jahall** [comment]
- _Efficient_ `K_diag` - Done

**Comment by jahall** [comment]
@frgsimpson @awav All previous comments addressed - do you have any further thoughts?

File: gpflow/kernels/changepoints.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            raise ValueError("Number of kernels ({nk}) must be one more than the number of "
+                             "changepoint locations ({nl})".format(nk=len(kernels), nl=len(locations)))
+
+        if isinstance(steepness, Iterable) and len(steepness) != len(locations):
+            raise ValueError("Dimension of steepness ({ns}) does not match number of changepoint "
+                             "locations ({nl})".format(ns=len(steepness), nl=len(locations)))
+
+        super().__init__(kernels, name=name)
+
+        self.num_changepoints = len(locations)
```
[the remark this answers is not in the document]
Comment (jahall) [reply gpflow/kernels/changepoints.py]: I felt that having the attribute made the code slightly more readable...but I can achieve the same by simply assigning a variable. Will change

**Comment by jahall** [comment]
Thanks for review @awav - I have resolved those comments. I've also added in an illustrative notebook to demo the kernel (had forgotten to `git add` this back in when I moved it to the new notebook location.

File: tests/test_kernels.py [hunk 2]
```diff
@@ -77,6 +77,25 @@ def _ref_periodic(X, lengthScale, signal_variance, period):
     return signal_variance * exp_dist
 
 
+def _ref_changepoints(X, kernels, locations, steepness):
```
[the remark this answers is not in the document]
Comment (jahall) [reply tests/test_kernels.py]: Sure thing

**Comment by awav** [comment]
Hello @jahall, I apologize for closing your PR. It was closed automatically when I moved `develop-2.0` to the `develop`. Could you re-open the PR? Thanks!

**Comment by jahall** [comment]
@awav @st-- Just wondering if you think this PR (and #1088) needs any further work.

File: tests/test_kernels.py [hunk 3]
```diff
@@ -77,6 +77,28 @@ def _ref_periodic(X, lengthScale, signal_variance, period):
...
+    # Calculates K(X) for each kernel in `kernels`, then multiply by sigmoid functions
```
Comment (awav) [inline tests/test_kernels.py]: Can you make it as a documentation with """...""". Thanks!

File: tests/test_kernels.py [hunk 4]
```diff
@@ -178,13 +200,13 @@ def test_periodic_1d_and_2d(D, N, lengthscale, variance, period):
...
 
 kernel_setups = [
-                    kernel() for kernel in gpflow.kernels.Stationary.__subclasses__()
-                ] + [
-                    gpflow.kernels.Constant(),
-                    gpflow.kernels.Linear(),
-                    gpflow.kernels.Polynomial(),
-                    gpflow.kernels.ArcCosine()
-                ]
+    kernel() for kernel in gpflow.kernels.Stationary.__subclasses__()
```
Comment (awav) [inline tests/test_kernels.py]: As far as I understand that this change was caused by your IDE? Is that right?

File: gpflow/kernels/changepoints.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        super().__init__(kernels, name=name)
+
+        self.locations = Parameter(locations)
+        self.steepness = Parameter(steepness, transform=positive())
+
+    def _set_kernels(self, kernels):
+        # it is not clear how to flatten out nested change-points
+        self.kernels = list(kernels)
+
+    def K(self, X, X2=None, presliced=False):
```
Comment (awav) [inline gpflow/kernels/changepoints.py]:
```suggestion
    def K(self, X: tf.Tensor, X2: Optional[tf.Tensor] = None, presliced: bool = False) -> tf.Tensor:
```

^^^ this also might require to import `Optional` from typing

File: gpflow/kernels/changepoints.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        N = tf.shape(X)[0]
+        ones = tf.ones((N, N, 1), dtype=X.dtype)
+        starters = tf.concat([ones, starters], axis=2)
+        stoppers = tf.concat([stoppers, ones], axis=2)
+
+        # now combine with the underlying kernels
+        kernel_stack = tf.stack([k(X, X2) for k in self.kernels], axis=2)
+        return tf.reduce_sum(kernel_stack * starters * stoppers, axis=2)
+
+    def K_diag(self, X, presliced=False):
```
Comment (awav) [inline gpflow/kernels/changepoints.py]:
```suggestion
    def K_diag(self, X: tf.Tensor, presliced: bool = False) -> tf.Tensor:
```

File: gpflow/kernels/changepoints.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        sig_X = tf.reshape(self._sigmoids(X), (N, -1))  # N x Ncp
+
+        ones = tf.ones((N, 1), dtype=X.dtype)
+        starters = tf.concat([ones, sig_X * sig_X], axis=1)  # N x Ncp
+        stoppers = tf.concat([(1 - sig_X) * (1 - sig_X), ones], axis=1)
+
+        kernel_stack = tf.stack([k(X, full=False) for k in self.kernels], axis=1)
+        return tf.reduce_sum(kernel_stack * starters * stoppers, axis=1)
+
+    def _sigmoids(self, X):
```
Comment (awav) [inline gpflow/kernels/changepoints.py]:
```suggestion
    def _sigmoids(self, X: tf.Tensor) -> tf.Tensor:
```

File: gpflow/kernels/base.py [hunk 8]
```diff
@@ -158,6 +158,9 @@ def __init__(self, kernels, name=None):
             raise TypeError(
                 "can only combine Kernel instances")  # pragma: no cover
 
+        self._set_kernels(kernels)
+
+    def _set_kernels(self, kernels):
```
Comment (awav) [inline gpflow/kernels/base.py]:
```suggestion
    def _set_kernels(self, kernels: List[Kernel]):
```

^^^ might require import List, or you can use Iteratable (if it allows to specify the element type)

**Comment by awav** [comment]
LGTM, please address comments and I'm okay with merging it after. Good work, thanks @jahall!

File: tests/test_kernels.py (hunk 3 above)
Comment (jahall) [reply tests/test_kernels.py]: Done

File: tests/test_kernels.py (hunk 4 above)
Comment (jahall) [reply tests/test_kernels.py]: No, I manually removed the superfluous space to be in keeping with the other `kernel_setups_*`

**Comment by jahall** [comment]
> LGTM, please address comments and I'm okay with merging it after. Good work, thanks @jahall!

Thanks @awav, have made appropriate changes.

I'm also keen to merge my generalized periodic kernel to v2.0 (see #1052 for v1.0 implementation). However, #1088 is a necessary first step for that, so am also interested in your feedback on that.
